src/drem/modelling/res_build_stock_postcode.py

```diff
@@ -0,0 +1,309 @@
+using the Split-Apply-Combine pandas transform. These ratios will
+then be applied to outputs from EnergyPlus to generate a first-pass
+estimate for residential energy demand in Dublin.
+
+"""
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+
+from prefect import Flow
+from prefect import task
+
+from drem.filepaths import PROCESSED_DIR
+from drem.filepaths import RAW_DIR
+from drem.filepaths import ROUGHWORK_DIR
+
+
+@task
+def _read_sa_parquet(input_filepath: str) -> pd.DataFrame:
+
+    return pd.read_parquet(input_filepath).drop_duplicates()
+
+
+@task
+def _read_sa_geometries(input_filepath: str) -> gpd.GeoDataFrame:
+
+    return gpd.read_parquet(input_filepath)
+
+
+@task
+def _read_csv(input_filepath: str) -> pd.DataFrame:
+
+    return pd.read_csv(input_filepath, encoding="unicode_escape").drop_duplicates()
+
+
+@task
+def _read_energy_csv(input_filepath: str) -> pd.DataFrame:
+
+    return pd.read_csv(input_filepath)
+
+
+@task
+def _extract_ber_dublin(df: pd.DataFrame, on: str, contains: str) -> pd.DataFrame:
+
+    return df[df[on].str.contains(contains)]
+
+
+@task
+def _lowercase_postcode(df: pd.DataFrame, new: str, old: str) -> pd.DataFrame:
+
+    df[new] = df[old].str.lower()
+
+    return df
+
+
+@task
+def _merge_ber_sa(
+    left: pd.DataFrame, right: pd.DataFrame, left_on: str, right_on: str, **kwargs,
+) -> pd.DataFrame:
+
+    return left.merge(right, left_on=left_on, right_on=right_on, **kwargs)
+
+
+@task
+def _split_ber_construction(df: pd.DataFrame, condition: str) -> pd.DataFrame:
+
+    return df.query(condition).drop_duplicates()
+
+
+@task
+def _extract_res(df: pd.DataFrame, on: str, value: str) -> pd.DataFrame:
+
+    return df.loc[df[on] == value].drop_duplicates().reset_index(drop=True)
+
+
+@task
+def _transform_res(df: pd.DataFrame, lon: str, lat: str) -> gpd.GeoDataFrame:
+
+    return gpd.GeoDataFrame(df, geometry=gpd.points_from_xy(lon, lat))
+
+
+@task
+def _spatial_join(
+    left: gpd.GeoDataFrame, right: gpd.GeoDataFrame, **kwargs,
+) -> gpd.GeoDataFrame:
+
+    right = right.set_crs(epsg="4326")
+
+    return gpd.sjoin(left, right)
+
+
+@task
+def _assign_building_type(df: pd.DataFrame, on: str, equiv: list) -> pd.DataFrame:
+
+    return df.replace({on: equiv})
+
+
+@task
+def _total_res_buildings_by_sa(
+    gdf: gpd.GeoDataFrame, on: str, renamed: str,
+) -> pd.DataFrame:
+
+    return gdf[[on]].value_counts().rename(renamed).reset_index()
+
+
+@task
+def _apply_period_built_split(
+    df: pd.DataFrame,
+    sa: str,
+    yc: str,
+    split: str,
+    year: int,
+    post: str,
+    pre: str,
+    pre_post: str,
+) -> pd.DataFrame:
+
+    df = df.groupby([sa, yc])
+    df = df.head(300000)
+    df[yc] = df[yc].astype(int)
+    df[split] = np.where(df[yc] >= year, post, pre)
+    df = df.groupby(sa)[split].value_counts(normalize=True)
+    df = df.to_frame()
+    df = df.rename(columns={split: pre_post})
+
+    return df.reset_index()
```

Comment:
Variable names could be improved, use year_of_construction instead of yc etc.

Implementation could be improved by using pandas built-in pd.DataFrame.where or pd.DataFrame.mask and so avoiding switching to numpy